### DISCUSSION TITLE
self._printChanges() causes infinite recursion

### DISCUSSION THREAD

**Discussion by iharandreyev** [pointfreeco/swift-composable-architecture#3500] [Q&A] [answered]
Hello everybody.
I am experimenting with SwiftUI navigation and TCA.
And I've stumbled upon a weird case when using _printChanges() on root reducer causes app to crash upon sending an action.
Stack trace shows that there are a lot of things being invoked again and again.

I'm pretty sure I've messed up something with Scoping.
Please help.

TCA is 1.15.2
Xcode 16.1
Target IOS is 16.0

The project is attached:
[ScreensAppNative.zip](https://github.com/user-attachments/files/17784550/ScreensAppNative.zip)

**Comment by mbrandonw** [comment, marked as answer]
Hi @iharandreyev, you cannot use `_printChanges` in this way:

```swift
var body: some ReducerOf<Self> {
  self._printChanges()
  …
}
```

That is composing a reducer into itself, and hence the infinite loop. The same thing would happen in vanilla SwiftUI if you did this:

```swift
var body: some View {
  self
}
```

Instead you need to apply `_printChanges()` to a reducer, like at the root:

```swift
let rootStore = Store(initialState: Root.State()) {
  Root()
    ._printChanges()
}
```
